### ISSUE TITLE
file API issues (specifically downloading)

### ISSUE THREAD

**Issue by finger563** [Kinvey/nativescript-sdk#36]
1. The Nativescript Files API / Guide listed on the Kinvey website is not representative of the actual API provided by this plugin
2. The returned type of the file data when using `Kinvey.Files.download('<file id>')` is not file metadata with a `_data` attribute as listed on the guide, but is actually the raw data for the file as an _encoded string_.  I'm actually having a good bit of difficulty right now trying to recover the actual binary data (which is what the file actually stores) from this string representation. I have checked to confirm that the file in the file store actually contains the right data - I have confirmed this by downloading the file through the kinvey developer console and checking the bytes of the downloaded file. 

Can someone please point me to a method for retrieving the binary data that I've downloaded when using the kinvey nativescript SDK? right now any time i'm trying to convert from any known encoding the data is corrupted / lost (meaning some bytes are correct, but multiple different byte pairs are all converted to unicode 65533.

I'm doing this right now in android on a device running android 7.0 and I have tried using `download` and `downloadByUrl` which both give me the exact same data (when using the `'<file id>'` or `'<file url>'` respectively, of course).

I am using the latest SDK - version 3.11.1.

**Comment by finger563** [comment]
e.g. here is the actual first few bytes of the file that I'm trying to download:
```
12 63 fe ec 00 00 fc ff ff ff ff
```
and here are the char codes (in hex form) that i've received directly from the downloaded data by using `data.charCodeAt(i).toString(16)`):
```
12 63 fffd fffd 0 0 fffd fffd fffd fffd fffd
```

As you can see - i've gotten some bytes correctly but other bytes are not

**Comment by finger563** [comment]
For reference, the `mime-type` is `application/octet-stream` but the `file` returned by `download()` is always of type `string`.

**Comment by finger563** [comment]
for now i've bypassed it and am just calling `httpModule.getFile( '<file._downloadURL>')`

**Comment by thomasconner** [comment]
@finger563 The data returned is either [NSData](https://developer.apple.com/documentation/foundation/nsdata) if the application is run on iOS or a [ByteArrayOutputStream](https://developer.android.com/reference/java/io/ByteArrayOutputStream). You should be able to use these data types and convert them to an image. For android you would do the following:

```java
ByteArrayOutputStream blob = new ByteArrayOutputStream();
byte[] bitmapdata = blob.toByteArray();
Bitmap bitmap = BitmapFactory.decodeByteArray(bitmapdata, 0, bitmapdata.length);
```
_I have not tested the above code but took the example from https://stackoverflow.com/questions/7620401/how-to-convert-byte-array-to-bitmap._

**Comment by thomasconner** [comment]
Can you see if this works for you?

**Comment by finger563** [comment]
it's not though - it's always a `string` - i have tried those methods before (as they're what i use when I load files from the FS for instance)

**Comment by thomasconner** [comment]
Can you provide an example app that returns to downloaded file as a string? I was testing file downloads a few days ago and did not receive them as a string but as the types mentioned above.  I haven’t been able to reproduce the issue you describe.

**Comment by finger563** [comment]
sure - it'll take a little bit to make a simple version of the app in question though.  while I do that - are you using `Angular` as well? And what is the `mime-type` of the file you're downloading?

**Comment by thomasconner** [comment]
@finger563 this has been resolved with 3.12.0 that was released today. On iOS you will be returned the file content as `NSData` and on Android it will be returned as an `ByteArrayOutputStream` when downloading a file. We updated our docs for files with this release. You can read the new documentation at https://devcenter.kinvey.com/nativescript/guides/files#Download.
